### ISSUE TITLE
Adopting a WebGL library for the internal WebGL implementation

### ISSUE THREAD

**Issue by trusktr** [trusktr/infamous#135]
It was great learning how to write WebGL from scratch, but it's lots of work and many problems have already been solved that we don't need to solve all over again.

Infamous gives end developers an interface for writing 3D scenes declaratively (using HTML) or imperatively (using JavaScript). This interface doesn't need to change. But under the hood, the way we render WebGL can be implemented with a WebGL library (as an example, [A-Frame](https://aframe.io) uses [Three.js](https://threejs.org) under the hood).

Three.js might be a bit heavy, more like a Framework with many features, while something like [Regl](http://regl.party) seems like the perfect foundation for starting a WebGL framework (for example, if I were writing a new Three.js from scratch, I'd start building on top of Regl).

A benefit of using Three.js would be that we get the hype that comes with Three.js, which can help us gain more traction. A benefit of using Regl would be a super clean and modular foundation that makes it easy to combine anything else written on top of Regl (f.e. if Three.js and [Babylon](http://www.babylonjs.com) were both written on top of Regl, it'd be super easy to render multiple things with both libraries _inside the same scene_).

---

Let's pick a library or libraries, and roll with it to make our first WebGL offering.

**Comment by trusktr** [comment]
I believe we'll go with Three.js. We'll also use this technique for "mixed mode": https://github.com/jeromeetienne/threex.htmlmixer

**Comment by trusktr** [comment]
I didn't understand what jeromeetienne/threex.htmlmixer was doing, so I made [my own](http://jsfiddle.net/trusktr/jc6j1wmf/) forked from [this one](http://jsfiddle.net/akmcv7Lh/56/) which shows how to line up DOM with WebGL planes. I took it further and lined up DOM with invisible planes, and used to invisible planes to "cut out" a sphere. This is how "Mixed Mode" can be done. It could also be used for physically mapped geometry in AR/VR/MR.

**Comment by trusktr** [comment]
I didn't understand what jeromeetienne/threex.htmlmixer was doing, so I made my own (http://jsfiddle.net/trusktr/jc6j1wmf), forked from this one (http://jsfiddle.net/akmcv7Lh/56) which shows how to align DOM elements with WebGL planes.

I took it further and lined up DOM elements with invisible WebGL planes, and used the invisible planes to "cut out" a sphere to give the illusion that the DOM elements intersect the sphere. This is how "Mixed Mode" can be done.

This technique can also be used with physically mapped geometry in AR/VR/MR, where, for example, your hand could intersect with a graphical object.
